File: app/src/main/java/ca/judacribz/week2weekend/adapters/CategoryAdapter.java
package ca.judacribz.week2weekend.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;

import ca.judacribz.week2weekend.R;
import ca.judacribz.week2weekend.models.Category;

public class CategoryAdapter extends ArrayAdapter<Category> {

    private Category category;

    public CategoryAdapter(Context context, ArrayList<Category> categories) {
        super(context, 0, categories);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        if (convertView == null) {
            convertView = LayoutInflater.from(getContext()).inflate(
                    R.layout.item_category,
                    parent,
                    false
            );
        }

        if ((category = getItem(position)) != null) {
            setText(convertView, R.id.tvCategoryName, category.getName());
            setText(convertView, R.id.tvNumSpecies, String.valueOf(category.getNumSpecies()));
            setText(convertView, R.id.tvDescription, category.getDescription());
        }

        return convertView;
    }

    private void setText(View convertView, int tvId, String name) {
        ((TextView) convertView.findViewById(tvId)).setText(name);
    }
}

File: app/src/main/java/ca/judacribz/zooatlanta/categories/CategoryAdapter.kt
package ca.judacribz.zooatlanta.categories

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import ca.judacribz.zooatlanta.R
import ca.judacribz.zooatlanta.models.Category
import kotlinx.android.synthetic.main.item_category.view.tvCategoryName
import kotlinx.android.synthetic.main.item_category.view.tvDescription
import kotlinx.android.synthetic.main.item_category.view.tvNumSpecies

class CategoryAdapter(private val categories: ArrayList<Category> = ArrayList(),
                      val onCategoryClick: (category: Category) -> Unit) :
    RecyclerView.Adapter<CategoryAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ViewHolder(LayoutInflater.from(parent.context).inflate(
            R.layout.item_category,
            parent,
            false
        ))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(categories[position])
    }

    override fun getItemCount() = categories.size

    fun addCategory(category: Category) {
        categories.add(category)
        notifyDataSetChanged()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(category: Category) = itemView.apply {
            setOnClickListener { onCategoryClick(category) }
            tvCategoryName.text = category.name
            tvNumSpecies.text = category.numSpecies.toString()
            tvDescription.text = category.description
        }
    }
}